lib/Sema/TypeCheckDecl.cpp

```diff
@@ -696,6 +696,165 @@ ExistentialTypeSupportedRequest::evaluate(Evaluator &evaluator,
 }
 
+static bool hasThrowingFunctionClosureParameter(CanType type) {
+  // Only consider throwing function types.
+  if (auto fnType = dyn_cast<AnyFunctionType>(type)) {
+    return fnType->getExtInfo().isThrowing();
+  }
+
+  // Look through tuples.
+  if (auto tuple = dyn_cast<TupleType>(type)) {
+    for (auto eltType : tuple.getElementTypes()) {
+      auto elt = eltType->lookThroughAllOptionalTypes()->getCanonicalType();
+      if (hasThrowingFunctionClosureParameter(elt))
+        return true;
+    }
+    return false;
+  }
+
+  // Suppress diagnostics in the presence of errors.
+  if (type->hasError()) {
+    return true;
+  }
+
+  return false;
+}
+
+static FunctionRethrowingKind
+getTypeThrowingKind(Type interfaceTy, GenericSignature genericSig) {
+  if (interfaceTy->isTypeParameter()) {
+    for (auto proto : genericSig->getRequiredProtocols(interfaceTy)) {
+      if (proto->isRethrowingProtocol()) {
+        return FunctionRethrowingKind::ByConformance;
+      }
+    }
+  } else if (auto NTD = interfaceTy->getNominalOrBoundGenericNominal()) {
+    if (auto genericSig = NTD->getGenericSignature()) {
+      for (auto req : genericSig->getRequirements()) {
+        if (req.getKind() == RequirementKind::Conformance) {
+          if (req.getSecondType()->castTo<ProtocolType>()
+                                 ->getDecl()
+                                 ->isRethrowingProtocol()) {
+            return FunctionRethrowingKind::ByConformance;
+          }
+        }
+      }
+    }
+  }
+  return FunctionRethrowingKind::Invalid;
+}
+
+static FunctionRethrowingKind 
+getParameterThrowingKind(AbstractFunctionDecl *decl, 
+                         GenericSignature genericSig) {
+  FunctionRethrowingKind kind = FunctionRethrowingKind::Invalid;
+  // check all parameters to determine if any are closures that throw
+  bool foundThrowingClosure = false;
+  for (auto param : *decl->getParameters()) {
+    auto interfaceTy = param->getInterfaceType();
+    if (hasThrowingFunctionClosureParameter(interfaceTy
+          ->lookThroughAllOptionalTypes()
+          ->getCanonicalType())) {
+      foundThrowingClosure = true;
+    }
+
+    if (kind == FunctionRethrowingKind::Invalid) {
+      kind = getTypeThrowingKind(interfaceTy, genericSig);
+    }
+  }
+  if (kind == FunctionRethrowingKind::Invalid &&
+      foundThrowingClosure) {
+    return FunctionRethrowingKind::ByClosure;
+  }
+  return kind;
+}
+
+ProtocolRethrowsRequirementList
+ProtocolRethrowsRequirementsRequest::evaluate(Evaluator &evaluator,
+                                              ProtocolDecl *decl) const {
+  SmallVector<std::pair<Type, ValueDecl*>, 2> found;
+  llvm::DenseSet<ProtocolDecl*> checkedProtocols;
+
+  ASTContext &ctx = decl->getASTContext();
+
+  // only allow rethrowing requirements to be determined from marked protocols
+  if (!decl->getAttrs().hasAttribute<swift::AtRethrowsAttr>()) {
+    return ProtocolRethrowsRequirementList(ctx.AllocateCopy(found));
```

Comment:
Isn't 'found' empty here? You can avoid the call to AllocateCopy and just return an empty ProtocolRethrowsRequirementList